### PR TITLE
Base project, few introductionary utilities

### PR DESCRIPTION
**Pull request by zvonicek**
- Support for CocoaPods, Carthage, SwiftPM
- iOS, MacOS, tvOS, watchOS targets
- Coordinator, SessionHolder, Nibable, Keyboardable, HairlineLayoutConstraint, UIView Extensions, proper UITableView/UICollectionView deselection handling

### CODE REVIEW

File: FuntastyKit.podspec [hunk 1]
```diff
@@ -0,0 +1,19 @@
+Pod::Spec.new do |s|
+  s.name         = "FuntastyKit"
+  s.version      = "0.1"
+  s.summary      = " A collection of Swift utilities and protocols used in our projects."
```
Comment (mkj-is) [inline FuntastyKit.podspec]: Space at the beginning of description?

File: FuntastyKit.podspec [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  s.description  = <<-DESC
+     A collection of Swift utilities and protocols used in our projects.
+  DESC
+  s.homepage     = "https://github.com/thefuntasty/FuntastyKit"
+  s.license      = { :type => "MIT", :file => "LICENSE" }
+  s.author             = { "Petr Zvonicek" => "" }
+  s.social_media_url   = ""
```
Comment (mkj-is) [inline FuntastyKit.podspec]: Funtasty Twitter?

File: Sources/Architecture/Coordinator.swift [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+            navigationController.present(destinationNavigationController, animated: animated, completion: nil)
+        } else {
+            // no wrapper navigation controller given, present actual controller
+            navigationController.present(viewController, animated: animated, completion: nil)
+        }
+    }
+
+    func stop() {
+        delegate?.willStop(in: self)
+        viewController?.dismiss(animated: true, completion: {
```
Comment (mkj-is) [inline Sources/Architecture/Coordinator.swift]: What about?

`viewController?.dismiss(animated: true) {`

File: Sources/Architecture/Coordinator.swift [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+            // present controller normally (initializer for this case not implemented, just an exploration of a possible future case)
+            navigationController?.pushViewController(viewController, animated: animated)
+        }
+    }
+
+    func stop() {
+        delegate?.willStop(in: self)
+
+        // TODO: figure out better way to distinguish between Push and Modal behavior
+        if let _ = destinationNavigationController {
```
Comment (mkj-is) [inline Sources/Architecture/Coordinator.swift]: I propose `!= nil`

File: Sources/Architecture/Coordinator.swift [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+            viewController?.dismiss(animated: true, completion: {
```
Comment (mkj-is) [inline Sources/Architecture/Coordinator.swift]: Completion block?

File: Sources/Extensions/UIWindow+RootViewController.swift [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        if animated {
+
+            UIView.transition(with: self, duration: 0.3, options: .transitionCrossDissolve, animations: { () -> Void in
+                let oldState = UIView.areAnimationsEnabled
+                UIView.setAnimationsEnabled(false)
+
+                self.rootViewController = viewController
+
+                UIView.setAnimationsEnabled(oldState)
+            }, completion: { (finished) -> Void in
```
Comment (mkj-is) [inline Sources/Extensions/UIWindow+RootViewController.swift]: Maybe improve the completion and send the bool from the function paramater.

File: Sources/Extensions/UIWindow+RootViewController.swift [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+     Set the rootViewController on this UIWindow instance.
+
+     - parameter viewController: The view controller to set
+     - parameter animated:       Whether or not to animate the transition, animation is a cross-fade
+     - parameter completion:     Completion block to be invoked after the transition finishes
+     */
+    @nonobjc func setRootViewController(_ viewController: UIViewController, animated: Bool, completion: @escaping () -> Void = {}) {
+        if animated {
+
+            UIView.transition(with: self, duration: 0.3, options: .transitionCrossDissolve, animations: { () -> Void in
```
Comment (mkj-is) [inline Sources/Extensions/UIWindow+RootViewController.swift:25]: Maybe add the durartion to function paramater with default value.

File: Sources/FuntastyKit.swift [hunk 8]
```diff
@@ -0,0 +1,25 @@
+/**
+ *  FuntastyKit
```
Comment (mkj-is) [inline Sources/FuntastyKit.swift:2]: 👍

**Review by mkj-is — CHANGES_REQUESTED** [review]
The most important thing is to review all the methods, protocols and objects if they are correctly accessible from outside modules.

File: Sources/Extensions/UIWindow+RootViewController.swift (hunk 6 above)
Comment (zvonicek) [reply Sources/Extensions/UIWindow+RootViewController.swift]: + pass nil value there if needed

File: FuntastyKit.podspec [hunk 9]
```diff
@@ -1,14 +1,14 @@
 Pod::Spec.new do |s|
   s.name         = "FuntastyKit"
```
Comment (hrach) [inline FuntastyKit.podspec]: the random indentation of `=` is not very nice :-)

Comment (zvonicek) [reply FuntastyKit.podspec]: I don't like it either, this file was actually pre-generated ;-)

**Review by mkj-is — APPROVED** [review]

**Pull request merged**